internal/characters/qiqi/abil.go

```diff
@@ -0,0 +1,219 @@
+		)
+		d.Targets = core.TargetAll
+
+		// One healing proc happens immediately on cast
+		c.Core.Health.HealActive(c.Index, c.heal(&d, skillHealContPer, skillHealContFlat, c.TalentLvlSkill()))
+
+		return &d
+	}, f)
+
+	// Queue up continuous healing instances
+	// No exact frame data on when the healing ticks happen. Just roughly guessing here
+	// Healing ticks happen 3 additional times during the skill - assume ticks are roughly 4.5s apart
+	// so in sec (0 = skill cast), 1, 5.5, 10, 14.5
+	c.AddTask(c.skillHealTickTask(src), "qiqi-skill-heal-tick", f+4.5*60)
+
+	// Queue up damage swipe instances.
+	// No exact frame data on when the damage ticks happen. Just roughly guessing here
+	// Occurs 9 times over the course of the skill
+	// Once shortly after initial cast, then 8 additional procs over the rest of the duration
+	// Each proc occurs in "pairs" of two swipes each spaced around 2.25s apart
+	// The time between each swipe in a pair is about 1s
+	// No exact frame data available plus the skill duration is affected by hitlag
+	// Damage procs occur (in sec 0 = skill cast): 1.5, 3.75, 4.75, 7, 8, 10.25, 11.25, 13.5, 14.5
+	c.AddTask(c.skillDmgTickTask(src, 60), "qiqi-skill-dmg-tick", f+30)
+
+	c.SetCD(core.ActionSkill, 30*60)
+
+	return f, a
+}
+
+// Handles skill damage swipe instances
+// Also handles C1:
+// When the Herald of Frost hits an opponent marked by a Fortune-Preserving Talisman, Qiqi regenerates 2 Energy.
+func (c *char) skillDmgTickTask(src int, lastTickDuration int) func() {
+	return func() {
+		if c.Core.Status.Duration("qiqiskill") == 0 {
+			return
+		}
+
+		// TODO: Not sure how this interacts with sac sword... Treat it as only one instance can be up at a time for now
+		if c.skillLastUsed > src {
+			return
+		}
+
+		d := c.Snapshot(
+			"Herald of Frost: Skill Damage",
+			core.AttackTagElementalArt,
+			core.ICDTagElementalArt,
+			core.ICDGroupDefault,
+			core.StrikeTypeDefault,
+			core.Cryo,
+			25,
+			skillDmgCont[c.TalentLvlSkill()],
+		)
+		d.Targets = core.TargetAll
+
+		if c.Base.Cons >= 1 {
+			d.OnHitCallback = c.c1
+		}
+
+		c.Core.Combat.ApplyDamage(&d)
+
+		nextTick := 60
+		if lastTickDuration == 60 {
+			nextTick = 135
+		}
+		c.AddTask(c.skillDmgTickTask(src, nextTick), "qiqi-skill-dmg-tick", nextTick)
+	}
+}
+
+func (c *char) c1(t core.Target) {
+	if c.talismanExpiry[t.Index()] < c.Core.F {
+		return
+	}
+	c.AddEnergy(2)
+
+	c.Core.Log.Debugw("Qiqi C1 Activation - Adding 2 energy", "frame", c.Core.F, "event", core.LogCharacterEvent, "char", c.Index, "target", t.Index())
+}
+
+// Handles skill auto healing ticks
+func (c *char) skillHealTickTask(src int) func() {
+	return func() {
+		if c.Core.Status.Duration("qiqiskill") == 0 {
+			return
+		}
+
+		// TODO: Not sure how this interacts with sac sword... Treat it as only one instance can be up at a time for now
+		if c.skillLastUsed > src {
+			return
+		}
+
+		// Create a temporary snapshot to get the atk mods included
```

Comment:
Is there a reason why you didn't use this function to get ATKP and ATK instead of getting a full snapshot? https://github.com/genshinsim/gcsim/blob/0a1394e66d5b675ad6ba6ebbbe0dacf2612c87df/pkg/character/stats.go#L28